org.adempiere.ui.zk/WEB-INF/src/org/adempiere/webui/window/WPAttributeDialog.java

```diff
@@ -535,140 +534,176 @@ private void addAttributeLine (Rows rows, MAttribute attribute, boolean product,
 		
 		m_row++;
-		Label label = new Label (attribute.getName());
-		if (product)
-			label.setStyle("font-weight: bold");
-			
-		if (attribute.getDescription() != null)
-			label.setTooltiptext(attribute.getDescription());
-		
-		Row row = rows.newRow();
-		row.appendChild(label.rightAlign());
+		WEditor editor = null;
 		//
 		
 		if (MAttribute.ATTRIBUTEVALUETYPE_List.equals(attribute.getAttributeValueType()))
 		{
-			MAttributeValue[] values = attribute.getMAttributeValues();	//	optional = null
-			Listbox editor = new Listbox();
-			editor.setMold("select");
-			for (MAttributeValue value : values) 
-			{
-				ListItem item = new ListItem(value != null ? value.getName() : "", value);
-				editor.appendChild(item);
-			}
-			if (readOnly)
-				editor.setEnabled(false);
-			else
-				m_editors.add (editor);
-			row.appendChild(editor);
-			ZKUpdateUtil.setHflex(editor, "1");
-			setListAttribute(attribute, editor);
+			editor = WebEditorFactory.getEditor(getListTypeGridField(attribute), true);
 		}
 		else if (MAttribute.ATTRIBUTEVALUETYPE_Number.equals(attribute.getAttributeValueType()))
 		{
-			NumberBox editor = new NumberBox(false);
-			setNumberAttribute(attribute, editor);
-			row.appendChild(editor);
-			ZKUpdateUtil.setHflex(editor, "1");
-			if (readOnly)
-				editor.setEnabled(false);
-			else
-				m_editors.add (editor);
+			editor = WebEditorFactory.getEditor(getNumberGridField(attribute), true);
 		}
-		else if(MAttribute.ATTRIBUTEVALUETYPE_Date.equals(attribute.getAttributeValueType()))
+		else if (MAttribute.ATTRIBUTEVALUETYPE_Reference.equals(attribute.getAttributeValueType()))
 		{
-			Datebox editor = new Datebox();
-			setDateAttribute(attribute, editor);
-			row.appendChild(editor);
-			if(readOnly)
-				editor.setEnabled(false);
-			else
-				m_editors.add(editor);
-			
+			editor = WebEditorFactory.getEditor(getGridField(attribute), true);
+		}
+		else
+		// Text Field
+		{
+			editor = WebEditorFactory.getEditor(getStringGridField(attribute), true);
 		}
-		else	//	Text Field
+
+		if (editor != null)
 		{
-			Textbox editor = new Textbox();
-			setStringAttribute(attribute, editor);
-			row.appendChild(editor);
-			ZKUpdateUtil.setHflex(editor, "1");
+			Row row = rows.newRow();
+
+			Label label = editor.getLabel();
+			if (label.getValue() == null || label.getValue().trim().length() < 1)
+				label.setValue(attribute.getName());
+
+			if (product)
+				label.setStyle("font-weight: bold");
+
+			row.appendChild(label.rightAlign());
+
+			editor.setMandatory(attribute.isMandatory());
+			editor.fillHorizontal();
+			setEditorAttribute(attribute, editor);
+
+			Component fieldEditor = editor.getComponent();
+			row.appendChild(fieldEditor);
+			editor.showMenu();
 			if (readOnly)
-				editor.setEnabled(false);
+				editor.setReadWrite(false);
 			else
-				m_editors.add (editor);
+				m_editors.add(editor);
 		}
+//		else if(MAttribute.ATTRIBUTEVALUETYPE_Date.equals(attribute.getAttributeValueType()))

```

Comment:
This block still needs to be translated and moved to the proper place. The date attribute type was not available when the modification was first introduced.So the original patch file does not deal with that.